includes/REST_API/Embed_Controller.php

```diff
@@ -0,0 +1,320 @@
+			[
+				[
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => [ $this, 'get_proxy_item' ],
+					'permission_callback' => [ $this, 'get_proxy_item_permissions_check' ],
+					'args'                => [
+						'url' => [
+							'description' => __( 'The URL for which to fetch embed data.', 'web-stories' ),
+							'type'        => 'string',
+							'format'      => 'uri',
+							'required'    => true,
+						],
+					],
+				],
+			]
+		);
+	}
+
+	/**
+	 * Callback for the Web Stories embed endpoints.
+	 *
+	 * Returns information about the given story.
+	 *
+	 * @param WP_REST_Request $request Full data about the request.
+	 *
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function get_proxy_item( $request ) {
+		$url = untrailingslashit( $request['url'] );
+
+		if ( empty( $url ) ) {
+			return new WP_Error( 'rest_invalid_url', __( 'Invalid URL', 'web-stories' ), [ 'status' => 404 ] );
+		}
+
+		/**
+		 * Filters the link data TTL value.
+		 *
+		 * @param int $time Time to live (in seconds). Default is 1 day.
+		 * @param string $url The attempted URL.
+		 */
+		$cache_ttl = apply_filters( 'web_stories_embed_data_cache_ttl', DAY_IN_SECONDS, $url );
+		$cache_key = 'web_stories_embed_data_' . md5( $url );
+
+		$data = get_transient( $cache_key );
+		if ( ! empty( $data ) ) {
+			if ( 'rest_invalid_story' === $data ) {
+				return new WP_Error( 'rest_invalid_story', __( 'URL is not a story', 'web-stories' ), [ 'status' => 404 ] );
+			}
+
+			return rest_ensure_response( json_decode( $data, true ) );
+		}
+
+		$data = $this->get_data_from_post( $url );
+		if ( $data ) {
+			return rest_ensure_response( $data );
+		}
+
+		$args = [
+			'limit_response_size' => 153600, // 150 KB.
+			'timeout'             => 7, // phpcs:ignore WordPressVIPMinimum.Performance.RemoteRequestTimeout.timeout_timeout
+		];
+
+		/**
+		 * Filters the HTTP request args for link data retrieval.
+		 *
+		 * Can be used to adjust timeout and response size limit.
+		 *
+		 * @param array $args Arguments used for the HTTP request
+		 * @param string $url The attempted URL.
+		 */
+		$args = apply_filters( 'web_stories_embed_data_request_args', $args, $url );
+
+		$response = wp_safe_remote_get( $url, $args );
+
+		if ( WP_Http::OK !== wp_remote_retrieve_response_code( $response ) ) {
+			// Not saving the error response to cache since the error might be temporary.
+			return new WP_Error( 'rest_invalid_url', __( 'Invalid URL', 'web-stories' ), [ 'status' => 404 ] );
+		}
+
+		$html = wp_remote_retrieve_body( $response );
+
+		if ( ! $html ) {
+			set_transient( $cache_key, 'rest_invalid_story', $cache_ttl );
+			return new WP_Error( 'rest_invalid_story', __( 'URL is not a story', 'web-stories' ), [ 'status' => 404 ] );
+		}
+
+		$data = $this->get_data_from_document( $html );
+
+		if ( ! $data ) {
+			set_transient( $cache_key, 'rest_invalid_story', $cache_ttl );
```

Comment:
Should we really be caching an error here? This is a bit of an anti pattern.

Reply:
No strong opinion here; removed caching errors now.